tools/syz-declextract/syz-declextract.cpp

```diff
@@ -64,6 +106,230 @@ bool beginsWith(const std::string_view &str, const std::string_view begin) {
+}
+
+const std::string makePtr(const std::string &dir, const std::string &type, bool isOpt = false) {
+  if (type == "void") {
+    return makePtr(dir, makeArray("int8"));
+  }
+  std::string ptr = "ptr[" + dir + ", " + type;
+  if (isOpt) {
+    return ptr + ", opt]";
+  }
+  return ptr + "]";
+}
+
+const std::string makeConst(const std::string &type, const std::string &val = "0") {
+  if (type.empty()) {
+    return "const[" + val + "]";
+  }
+  if (type == "void") {
+    return "const[" + val + ", intptr]";
+  }
+  return "const[" + val + ", " + type + "]";
+}
+
+const std::string getSyzType(const std::string &ctype, const bool isSyscallParam = false) {
+  if (!isSyscallParam && (contains(ctype, "long long") || contains(ctype, "64"))) {
+    return contains(ctype, "be") ? "int64be" : "int64";
+  }
+  if (contains(ctype, "16") || contains(ctype, "short")) {
+    return contains(ctype, "be") ? "int16be" : "int16";
+  }
+  if (contains(ctype, "8")) {
+    return contains(ctype, "be") ? "int8be" : "int8";
+  }
+  if (contains(ctype, "32") || contains(ctype, "int")) {
+    return contains(ctype, "be") ? "int32be" : "int32";
+  }
+  if (contains(ctype, "void")) {
+    return "void";
+  }
+  if (contains(ctype, "char")) {
+    return "int8";
+  }
+  return "intptr";
+}
+
+class RecordExtractor {
+private:
+  const SourceManager *const SM;
+  std::vector<std::string> includes;
+  std::vector<std::string> flags;
+  std::unordered_map<std::string, size_t> visitedRecord; // Maps name to index in extractedRecords
+  std::vector<SyzRecordDecl> extractedRecords;
+
+  unsigned int getCountedBy(const FieldDecl *const &field) {
+    return field->getType()->isCountAttributedType()
+               ? llvm::dyn_cast<FieldDecl>(
+                     field->getType()->getAs<CountAttributedType>()->getCountExpr()->getReferencedDeclOfCallee())
+                     ->getFieldIndex()
+               : UINT_MAX;
+  }
+
+  bool isFieldVarlen(const QualType &fieldType) {
+    return fieldType->isIncompleteArrayType() ||
+           (fieldType->isConstantArrayType() && llvm::dyn_cast<ConstantArrayType>(fieldType)->getSize().isZero());
+  }
+
+  std::string getStructAttr(const RecordDecl *const recordDecl, ASTContext *context) {
+    if (recordDecl->isStruct()) {
+      for (const auto &item : recordDecl->getAttrs()) {
+        if (item->getKind() == clang::attr::Aligned) {
+          return "align[" + std::to_string(llvm::dyn_cast<AlignedAttr>(item)->getAlignment(*context)) + "]";
+        } else if (item->getKind() == clang::attr::Packed) {
+          return "packed";
+        }
+      }
+    }
+    return "";
+  }
+
+public:
+  RecordExtractor(const SourceManager *const SM) : SM(SM){};
+  std::string getFieldType(const QualType &fieldType, ASTContext *context, const std::string &fieldName,
+                           const std::string &parent = "", bool isSyscallParam = false) {
+    if (contains(fieldName, "ifindex")) {
```

Comment:
I think this is a wrong place for this. Consider we have a pointer that happens to contain "ifindex" in the name. It's also wrong to use "ifindex" if the size of the field wasn't int32 b/c ifindex is int32, so if we use it we will break layout of the struct. 

A better way to do it is:
1. do it where we extract builtin int types (so no pointers).
2. do it after we extracted raw type (e.g. int32).
3. have heuristics function shared between this and netlink that will accept raw type + name and produce refined type if possible.
4. the function should take raw type into account.
5. the function should convert the name to lower case so that name heuristics work for netlink and field.